packages/flutter/lib/src/widgets/app.dart

```diff
@@ -677,33 +749,80 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
+    // be expensive to search through them many times.
+    Map<int, Locale> allSupportedLocales = HashMap<int, Locale>();
+    Map<int, Locale> languageAndCountryLocales = HashMap<int, Locale>();
+    Map<int, Locale> languageAndScriptLocales = HashMap<int, Locale>();
+    Map<int, Locale> languageLocales = HashMap<int, Locale>();
     for (Locale locale in supportedLocales) {
-      if (locale == newLocale)
-        return newLocale;
-      if (locale.languageCode == newLocale.languageCode)
-        matchesLanguageCode ??= locale;
+      allSupportedLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode, locale.countryCode.hashCode)] ??= locale;
+      languageAndCountryLocales[ui.hashValues(locale.languageCode.hashCode, locale.countryCode)] ??= locale;
+      languageAndScriptLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode)] ??= locale;
+      languageLocales[locale.languageCode.hashCode] ??= locale;
+    }
+    // Loop over user's preferred locales
+    Locale matchesLanguageCode;
+    for (int localeIndex = 0; localeIndex < preferredLocales.length; localeIndex++) {
+      Locale userLocale = preferredLocales[localeIndex];
+      // Look for perfect match.
+      if (allSupportedLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode, userLocale.countryCode.hashCode))) {
+        return userLocale;
+      }
+      // Wait until next userLocale to see if there is a high accuracy match before
+      // returning a language-only match.
+      if (matchesLanguageCode != null) {
+        return matchesLanguageCode;
+      }
+      // Look for language+country match.
+      if (languageAndCountryLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.countryCode.hashCode))) {
+        return languageAndCountryLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.countryCode.hashCode)];
+      }
+      // Look for language+script match.
+      if (languageAndCountryLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode))) {
+        return languageAndCountryLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode)];
+      }
+      // Look and store language only match.
+      if (languageLocales.containsKey(userLocale.languageCode.hashCode)) {
+        matchesLanguageCode = languageLocales[userLocale.languageCode.hashCode];
+        // Since first (default) locale is usually highly preferred, we will allow
+        // only a languageCode to suffice to be instantly matched.
+        if (localeIndex == 0) {
+          return matchesLanguageCode;
+        }
+      }
     }
     return matchesLanguageCode ?? supportedLocales.first;
   }
 
   @override
-  void didChangeLocale(Locale locale) {
-    if (locale == _locale)
+  void didChangeLocales(List<Locale> locales) {
+    if (locales.first == _locale)
```

Comment:
??

Reply:
This is the WIP part and I wanted to see how best to go about modifying these.